In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

### Pré traitement des données 

In [2]:
df = pd.read_csv('data/projects_data.csv')

In [3]:
df.head()

,project_name,link,description,composants
0,Tillu - the Robot,https://www.instructables.com/Tillu-the-Robot,"MeetTillu - The Robot, a unique fusion of adva...",1xUnihiker 4xServo 1xBattery Managment 2xUSB T...
1,Air Hockey Robot With Shot Prediction,https://www.instructables.com/Air-Hockey-Robot,Ever wanted to play air hockey against a compu...,Disclaimer - Supplies list may include affilia...
2,Two-Wheeled Drive Car With a Robotic Arm,https://www.instructables.com/Two-Wheeled-Driv...,This project was developed within theOrange Di...,To start building a Two-Wheeled Drive Car with...
3,Ghost Candy Dispenser,https://www.instructables.com/Ghost-Candy-Disp...,Hi there! Did you enjoy Halloween in 2024? Sin...,Parts MCU:M5StampS3 (M5Stack)(×1) RC servo mot...
4,Lily∞Bot With Micro:Bit and Motor:Bit Does Obs...,https://www.instructables.com/LilyBot-With-Mic...,This project will explain how to get obstacle ...,The parts list for the Lily∞Bot With Micro:Bit...


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1503 entries, 0 to 1502
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   project_name  1499 non-null   object
 1   link          1503 non-null   object
 2   description   1178 non-null   object
 3   composants    1169 non-null   object
dtypes: object(4)
memory usage: 47.1+ KB


In [5]:
df.isna().sum()

project_name      4
link              0
description     325
composants      334
dtype: int64

On a:
- *325* projets sans descriptions
- *334* projets dont les composants pour la concéption ne sont pas indiqués.

Je vais tous simplement supprimer les lignes sans l'un de ces attributs car on ne pourra pas récommender à une personne un projet si on ne sait même pas de quoi parle le projet...

In [6]:
df_cleaned = df.dropna(subset=['composants','description'])
df_cleaned.isnull().sum()

project_name    0
link            0
description     0
composants      0
dtype: int64

**Je vais traduire les textes en Français pour la suite de l'analyse**

In [7]:
from transformers import MarianMTModel, MarianTokenizer

/Users/amadouu/M2_ML/BD/env_bigData/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm

A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.1.3 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "/usr/local/Cellar/python@3.12/3.12.6/Frameworks/Python.framework/Versions/3.12/lib/python3.12/runpy.py", line 198, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/Cellar/python@3.12/3.12.6/Frameworks/Python.fr

In [10]:
from transformers import MBartForConditionalGeneration, MBart50TokenizerFast

model = MBartForConditionalGeneration.from_pretrained("facebook/mbart-large-50-many-to-many-mmt")
tokenizer = MBart50TokenizerFast.from_pretrained("facebook/mbart-large-50-many-to-many-mmt")


In [ ]:
def translate_text(text):
    translated = model.generate(**text)
    translated_text = tokenizer.decode(translated[0], skip_special_tokens=True)
    translated = model.generate(**tokenizer(src_text, return_tensors="pt", padding=True))
    [tokenizer.decode(t, skip_special_tokens=True) for t in translated] 
    return " ".join(translations)

In [131]:
df_cleaned['composants_fr'] = df_cleaned['composants'].apply(lambda x: translator.translate(x))

NotValidLength: The majority of our parts have been sourced from GoBilda, a proud materials provider for FTC robotics. Their parts are quite high quality, but also reasonably expensive. I won't provide every single structure, as there are too many individual pieces to count, but I'll list the major components, such as the control hub, driver and expansion hubs, and major pieces.  Off-the-Shelf Parts List: 1x REV Robotics Control Hub- $350.00 - https://www.revrobotics.com/rev-31-1595/ 1x REV Robotics Expansion Hub- $250.00 - https://www.revrobotics.com/rev-31-1153/ The Control and Expansion Hub is where all your programs and core information will be stored, as well as where all your wiring is connected to. 6x GoBilda Yellow Jacket 19:2:1 Gear Motors (Planetary, REX Bore)- $42.99/pc - https://www.gobilda.com/5203-series-yellow-jacket-planetary-gear-motor-19-2-1-ratio-24mm-length-8mm-rex-shaft-312-rpm-3-3-5v-encoder/ 6x GoBilda Yellow Jacket 5:2:1 Gear Motors (Planetary, REX Bore)- $42.99/pc - https://www.gobilda.com/5203-series-yellow-jacket-planetary-gear-motor-5-2-1-ratio-24mm-length-8mm-rex-shaft-1150-rpm-3-3-5v-encoder/ These DC motors work beautifully under heavy loads but make sure you choose the right torque and ratios. GoBilda Servos (Torque and Speed)- $31.99 - https://www.gobilda.com/standard-size-servos/ Servos are important for smaller attachments, such as a claw or smaller rollers. 12x Encoder Extension Cable- $2.99/pc - https://www.gobilda.com/encoder-cable-extension-4-pos-jst-xh-300mm-length/ While the drive motors have built-in encoders, it is important to purchase extensions for the motors in order to connect the drive encoders to the hubs. GoBilda U-Channel 1120 Series- Price Varies on Size - https://www.gobilda.com/1120-series-u-channel/ This link takes you to their U-channel page, where they have various sizes. This composes about half of our structure. They may be a little expensive, but the investment is worth it for us! Many of our channels are still going strong after 2-3 years. Misumi Telescopic Rails- Price ranges from $12.99 to $22.99 - https://us.misumi-ec.com/vona2/detail/110300072130/ We recommend Misumi telescopic rails for extension attachments, such as our extension arm. It is more of a DIY approach than other extendo rails such as GoBilda's, but with the high-quality build, it is COMPLETELY worth it in the long run. 96mm Mecanum Wheels (4-Set)- $169.99 - https://www.gobilda.com/96mm-mecanum-wheel-set-70a-durometer-bearing-supported-rollers/ Okay, I'll admit, $170 is A LOT for 4 mecanum wheels. However, you can use any wheelset as you'd please, as you can find sets for as low as 30 or 40 bucks on Amazon or any wholesaler. Screws and Fasteners- I'd recommend purchasing from GoBilda to keep the compatibility, but you can use anyM4set screws. Please ensure you are using M4, as all of these parts are based on M4 compatibility.  Custom Parts: Now approaching the topic of custom parts: for the most part, I used 3D printing to create parts such as telescopic slide inserts, ramps, and more, but you can CNC or laser-cut these as you prefer. For 3D printing, I recommend a PLA-based material for the majority of your prints, unless you have a high-impact scenario. PETG or ABS will work well in these cases. 3D Printing PLA Filament Overture- $17.99/KG - https://www.amazon.com/OVERTURE-Filament-Consumables-Dimensional-Accuracy/dp/B07PGY2JP1/ref=sr_1_4?crid=3TAD0QT7R75E3&keywords=pla&qid=1699764288&sprefix=pl%2Caps%2C116&sr=8-4  Tools and Software: I must admit here, I am not much of a software guy - I always had trouble learning software, and stuck with CAD as my lingo. However, I will still list software components to the best of my ability - apologize in advance if I miss anything. CAD:For the majority of this project, I utilized Fusion 360, an amazing tool to design all our components and assemble the robot in advance. 3D Printer:I would recommend a 3D printer with a bed size of at least 250mm x 250mm x 250mm. My Ender 3 Pro works amazing for this project, and if you want to make specialized parts (sorry - WHEN you need to make them... making custom parts areimportant), I'd recommend a decently sized printer. Android Studio- When coding the robot, you want to have a good coding studio, as you'll be making project after project of heavy coding, image recognition, tesselation analysis, odometry, and so much more. Allen Keys- Hex tools/Allen keys are crucial for these parts, as they come majority with hex head screws. Hammer- You'll need a hammer to smash things in place :) Filer/Sanding tool- This will come in handy for fixing the tolerances of your prints. Trust me when I say that shaving off a layer or two might come in handy, so have some sandpaper lying around!  Safety: I cannot emphasize enough how important safety is! From inhaling metal dust to laser cutting fumes to sawing polycarbonate, each and every action you complete has a risk in the lab. Here are a few things to remember at all times: Work in a well-ventilated area!Keep a window open, or have a vent pumping fresh air into your workspace, such as an AC or a fan. When cutting things, such as aluminum or polycarbonate, both of which can emit dangerous particles or gas, it is ideal to have avacuum pump or ventrunning into the cutting machine.A maskcan also suffice in case you don't have a pump machine. Safety gogglesandglovesare a MUST when working with parts!  There may be a lot to take in with this list of parts and actions, but it is important to keep all of this in mind; rather than this being a single, two to three-month-long project, this is a years-long project that can be worked on as a team. With the right funding and time set aside, you can create something truly amazing. Without further ado, let's jump into making the bot! --> Text length need to be between 0 and 5000 characters